electrolytes/solvation_shell_extract.py

```diff
@@ -0,0 +1,266 @@
+                ]
+                central_solute_nums = [mol.molecule_number for mol in solute_molecules]
+                # Extract solvation shells
+                shells = [
+                    set(
+                        analyze.evaluate_asl(
+                            st, f"fillres within {radius} mol {mol_num}"
+                        )
+                    )
+                    for mol_num in central_solute_nums
+                ]
+
+                # Now expand the shells
+                for shell, central_solute in zip(shells, central_solute_nums):
+                    expanded_shell = expand_shell(
+                        st,
+                        shell,
+                        central_solute,
+                        radius,
+                        solute_resnames,
+                        max_shell_size=200,
+                    )
+                    expanded_shells.append(expanded_shell)
+
+            # Now compare the expanded shells and group them by similarity
+            # we will get lists of lists of shells where each list of structures are conformers of each other
+            logging.info("Grouping solvation shells into conformers")
+            # TODO: speed this up
+            grouped_shells = group_with_comparison(expanded_shells, are_conformers)
+
+            # Now ensure that topologically related atoms are equivalently numbered (up to molecular symmetry)
+            grouped_shells = [
+                renumber_molecules_to_match(items) for items in grouped_shells
+            ]
+
+            # Now extract the top N most diverse shells from each group
+            logging.info(f"Extracting top {top_n} most diverse shells from each group")
+            final_shells = []
+            # example grouping - set of structures
+            for shell_group in tqdm(grouped_shells):
+                filtered = filter_by_rmsd(shell_group, n=top_n)
+                final_shells.extend(filtered)
+
+            # Save the final shells
+            logging.info(f"Saving final shells")
+            save_path = os.path.join(save_dir, system_name, species, f"radius={radius}")
+            os.makedirs(save_path, exist_ok=True)
+            for i, st in enumerate(final_shells):
+                # TODO: seems like this is saving an extra line at the end of the xyz files
+                st.write(os.path.join(save_path, f"shell_{i}.xyz"))
+
+    # Now repeat for solvents to capture solvent-solvent interactions
+    for species, residue in solvents.items():
+        logging.info(f"Extracting shells around {species}")
+        for radius in solvent_radii:
+            logging.info(f"Radius = {radius} A")
+            filtered_shells = []
+            for i, st in tqdm(enumerate(structures)):  # loop over timesteps
+                # assign partial charges to atoms
+                for at, charge in zip(st.atom, partial_charges):
+                    at.partial_charge = charge
```

Comment:
Is this important for some reason?

Reply:
I was under the impression that the final outputted files needed to include the partial charge information since it's needed for DFT, so I added it here. But it looks like it's not actually getting written to the output `.xyz` files properly (they still only include positions). Do you have a suggestion on how to do this? Maybe I need to write to PDB format to incorporate charge info?

Reply:
Final structures need the total charge and spin, not partial charges.

Reply:
Are these partial charges actually floats or are they formal charge? If you add all the charges to at.formal_charge then at the end when you have extracted things, you can do st.formal_charge to get the sum of formal charges on all the atoms.

Reply:
The partial charges are floats. But since we only need total charge for DFT, can we do the process you just described? I.e assign all partial charges, and get st.partial_charge at the end to get the sum across all atoms in a shell?